conans/client/rest/uploader_downloader.py

```diff
@@ -205,95 +161,57 @@ def _download_file(self, url, auth, headers, file_path):
-            if not file_path:
-                ret += response.content
-            else:
-                if self.output:
-                    total_length = len(response.content)
-                    if progress_bar is not None:
-                        progress_bar.desc = "Downloading {}".format(os.path.basename(file_path))
-                        progress_bar.total = total_length
-                        progress_bar.update(total_length)
-
-                save_append(file_path, response.content)
-        else:
+            total_length = response.headers.get('content-length') or len(response.content)
             total_length = int(total_length)
+            description = "Downloading {}".format(os.path.basename(file_path)) if file_path else None
+            progress = progress_bar.WriteProgress(total_length, self.output, description)
+
+            chunk_size = 1024 if not file_path else 1024 * 100
             encoding = response.headers.get('content-encoding')
             gzip = (encoding == "gzip")
-            # chunked can be a problem:
-            # https://www.greenbytes.de/tech/webdav/rfc2616.html#rfc.section.4.4
-            # It will not send content-length or should be ignored
-            if progress_bar is not None:
-                progress_bar.total = total_length
-
-            def download_chunks(file_handler=None, ret_buffer=None):
-                """Write to a buffer or to a file handler"""
-                chunk_size = 1024 if not file_path else 1024 * 100
-                download_size = 0
-                last_time = 0
-                if progress_bar is not None:
-                    progress_bar.desc = "Downloading {}".format(os.path.basename(file_path))
-
-                for data in response.iter_content(chunk_size):
-                    download_size += len(data)
-                    if ret_buffer is not None:
-                        ret_buffer.extend(data)
-                    if file_handler is not None:
-                        file_handler.write(to_file_bytes(data))
-                    if progress_bar is not None:
-                        progress_bar.update(len(data))
-                    elif self.output and time.time() - last_time > TIMEOUT_BEAT_SECONDS:
-                        last_time = time.time()
-                        self.output.write(TIMEOUT_BEAT_CHARACTER)
-
-                return download_size
-
-            if file_path:
-                mkdir(os.path.dirname(file_path))
-                with open(file_path, 'wb') as handle:
-                    dl_size = download_chunks(file_handler=handle)
-            else:
-                dl_size = download_chunks(ret_buffer=ret)
+            ret_data = bytearray()
 
-            response.close()
+            downloaded_chunks = written_chunks(
+                progress.update(read_response(chunk_size), chunk_size),
+                file_path
+            )
+
+            for data in downloaded_chunks:
```

Comment:
This seems weird. I'd say that this should be appended to ``ret_data`` in the ``written_chunks`` function, not here, to be symmetric with the write to disk. Rename to ``write_chunks`` and they will be written to disk or to memory.